src/main/java/com/google/firebase/database/util/EmulatorHelper.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.database.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.firebase.database.annotations.Nullable;
+import com.google.firebase.database.core.RepoInfo;
+import com.google.firebase.database.utilities.ParsedUrl;
+import com.google.firebase.database.utilities.Utilities;
+
+public final class EmulatorHelper {
+
+  private EmulatorHelper() {
+  }
+
+  @VisibleForTesting
+  public static final String FIREBASE_RTDB_EMULATOR_HOST_ENV_VAR = "FIREBASE_RTDB_EMULATOR_HOST";
+
+  public static String getEmulatorHostFromEnv() {
+    return System.getenv(FIREBASE_RTDB_EMULATOR_HOST_ENV_VAR);
+  }
+
+  @VisibleForTesting
+  @Nullable
+  static String tryToExtractEmulatorUrlFromDbUrl(String suppliedDatabaseUrl) {
+    if (Strings.isNullOrEmpty(suppliedDatabaseUrl)) {
+      return null;
+    }
+    ParsedUrl parsedUrl = Utilities.parseUrl(suppliedDatabaseUrl);
+    RepoInfo repoInfo = parsedUrl.repoInfo;
+    if (repoInfo.host.endsWith(".firebaseio.com") || !suppliedDatabaseUrl
+        .contains("ns=")) {
+      return null;
+    }
+    String pathString = parsedUrl.path.isEmpty() ? "/" : parsedUrl.path.toString() + "/";
+    String scheme = repoInfo.isSecure() ? "https" : "http";
+    return String.format("%s://%s%s?ns=%s", scheme, repoInfo.host, pathString, repoInfo.namespace);
+  }
+
+  public static String overwriteDatabaseUrlWithEmulatorHost(String suppliedDatabaseUrl,
+      String emulatorHost) {
+    String extractedEmulatorUrl = tryToExtractEmulatorUrlFromDbUrl(suppliedDatabaseUrl);
+    if (!Strings.isNullOrEmpty(extractedEmulatorUrl)) {
```

Comment:
Something still feels a little weird about the control flow here, especially since you are still mixing the concepts of databaseUrl and emulatorHost in somewhat surprising ways.

What do you think about this (pseudocode):

```
public static ParsedUrl getEmulatorUrl(String suppliedDatabaseUrl) {
    ParsedUrl parsedUrl = Utilities.parseUrl(suppliedDatabaseUrl);
    if (isEmulatorUrl(parsedUrl)) {
		return parsedUrl;
	}
	
	String hostEnv = getEmulatorHostFromEnv();
	if (!hostEnv) {
		return null;
	}
	
    ParsedUrl parsedUrl = Utilities.parseUrl(hostEnv);
    if (parsedUrl.scheme("http:") || parsedUrl.path("?ns=")) {
      throw new IllegalArgumentException(
          "emulator host declared in environment variable must be of the format \"host:port\"");
    }
	
    String namespaceName = "default";
    String path = "/";
    if (!Strings.isNullOrEmpty(suppliedDatabaseUrl)) {
      ParsedUrl parsedUrl = Utilities.parseUrl(suppliedDatabaseUrl);
      namespaceName = parsedUrl.repoInfo.namespace;
      path = parsedUrl.path.isEmpty() ? "/" : parsedUrl.path.toString() + "/";
    }
	
    return new ParsedUrl(...)
  }
```

Reply:
Thanks! Makes things much cleaner.
One caveat though: `isEmulatorUrl(String url)` has to take a string URL instead of `ParsedUrl`. This is because we need to check for the existence of the namespace query parameter.